massa-protocol-worker/src/connectivity.rs

```diff
@@ -310,35 +314,58 @@ pub(crate) fn start_connectivity_thread(
-                                                }
-                                            }
-                                        } else if slot_default_category > 0 &&  !addresses_to_connect.contains(addr) {
-                                            addresses_to_connect.push(*addr);
-                                            slot_default_category -= 1;
-                                        }
+                                        addresses_can_connect.push((*addr, connection_metadata, category_found));
+                                    }
+                                }
+                            }
+                        }
 
-                                            // IF all slots are filled, stop
-                                        if slot_default_category == 0 && slots_per_category.iter().all(|(_, slots)| *slots == 0) {
-                                            break;
+                        // Sort addresses using the metadata
+                        addresses_can_connect.sort_by(|a, b| a.1.cmp(&b.1));
+
+                        let mut addresses_to_connect = vec![];
+                        for (addr, _, category) in addresses_can_connect.iter() {
+                            match category {
+                                Some(cat) => {
+                                    for (name, category_infos) in &mut slots_per_category {
+                                        if name == *cat && category_infos > &mut 0 {
+                                            addresses_to_connect.push(*addr);
+                                            *category_infos -= 1;
                                         }
                                     }
                                 }
+                                None if !addresses_to_connect.contains(&addr) => {
+                                    slot_default_category -= 1;
+                                    addresses_to_connect.push(*addr);
+                                }
+                                _ => continue,
+                            }
+
+                            // IF all slots are filled, stop
+                            if slot_default_category == 0 && slots_per_category.iter().all(|(_, slots)| *slots == 0) {
+                                break;
                             }
                         }
 
                         for addr in addresses_to_connect {
                             info!("Trying to connect to addr {}", addr);
 
-                            peer_db.write().try_connect_history.insert(addr, MassaTime::now().unwrap());
+                            let conn_res = network_controller.try_connect(addr, config.timeout_connection.to_duration());
+                            {
+                                let mut peer_db_write = peer_db.write();
+                                let addr_conn_history = peer_db_write.try_connect_history.remove(&addr);    // Remove it to own it, then re-insert it modified
```

Comment:
why not get_mut() ? or try_connect_history.entry(XXX).and_modify(XXX)

Reply:
I'd honestly be open for a better way to handle this tbh